### PR TITLE
Fix dynamic imports using totally new import resolver

### PR DESCRIPTION
**Pull request by exogen**
This PR features a totally new import resolver for `stylus-relative-loader`. It has now outgrown its namesake, not only supporting correct relative imports but also **dynamic** imports (with variable interpolation, inside conditionals, etc.) That means we can very easily support multi-tenancy, e.g.

``` stylus
@require "~@walmart/lithe-extras/lib/tenants/" + $tenant + "/base";
```

The new import resolver passes the same test suite without modification. It works like so:
- Instead of trying to statically analyze imports in a first pass, the Stylus renderer is invoked immediately.
- Upon finding an import that can't be resolved (usually means it's a webpack import), a special flow-control exception is thrown. The exception provides information about the requested import to our "outer loop."
- Each unresolved import is sent to the `resolve` function webpack provides. If that succeeds we store information about the result and then simply try the whole thing again, reinvoking the renderer.

This works because unlike when bundling JavaScript, the purpose of this loader is to render Stylus files all the way to static CSS, effectively "running the program." So there is no need for a static analysis step; we can just watch what actual imports are attempted at runtime.

[This trickery (and also that done by `stylus-loader`) is all necessary because (1) Stylus doesn't expose a hook for its import resolver and (2) Stylus imports are synchronous, so even if there were an import hook, it wouldn't get along with webpack's async `resolve` function.]
### TODO
- [ ] Fix sourcemaps #4 
- [x] Add dynamic import tests #5 
- [ ] Make sure hot reloading works (the WML repo this was tested with seems to have broken HMR anyway) #6 
- [x] Better import caching / performance? #7
- [x] Set up `eslint`

/cc @ryanisinallofus

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -70,10 +76,12 @@ module.exports = function(source) {
         styl.define(defineName, value[defineName]);
       }
     } else if (key === 'include') {
-      needsArray(value).forEach(styl.include);
+      needsArray(value).forEach(function(path) {
```
Comment (ryan-roemer) [inline index.js]: NIT: Collision with `var path` import at top. Is eslint catching this?

Comment (exogen) [inline index.js]: Good catch; adding another TODO to set up `eslint`. Right now all the dev tooling is the same as upstream, no linting.

Side note about this particular line, it's fixing a bug in the upstream code. `styl.include` uses `this` (which should be the Stylus renderer instance) but this `forEach` wasn't passing `styl` as the context, so it's broken; apparently the `include` option just isn't tested.

We could potentially avoid the `path` naming here entirely by doing `needsArray(value).forEach(styl.include, styl)` instead, although that still relies on knowing that `include` only takes 1 parameter.

**Comment by exogen** [comment]
Lint added. Issues opened for all the rest. I'm comfortable with using a few applications as our testing ground for the new behavior for now. /cc @ryanisinallofus

**Comment by ryanisinallofus** [comment]
LGTM

**Comment by Kinrany** [comment]
Should the two open issues mentioned here be closed?

**Pull request merged**